arch/arm/soc/nordic_nrf5/nrf51/soc.c

```diff
@@ -34,6 +36,17 @@ static bool ftpan_59(void);
 
 uint32_t SystemCoreClock __used = __SYSTEM_CLOCK;
 
+void relocate_vector_table(void)
+{
+#if defined(CONFIG_IS_BOOTLOADER) || defined(CONFIG_IMG_MANAGER)
```

Comment:
There's no reason to connect vector relocation feature with CONFIG_IMG_MANAGER - it is different functionalities. image management itself won't use vector relocation. vector relocation call is needed for boot loader only.

Reply:
Except if you expect that vector relocation call should be call by runing application - then question is how application will know where placed bootloader's _ram_vector_start[] is?

Reply:
Well, I find zephyr has three way to boot.
And I design it like this:
1).Boot as a bootloader
Boot as bootloader, zephyr must have a vector relay function put at address 0x00000000, we use 0x20000000 as proxy to relay irq. And application and bootloader it self must copy its own vector table to 0x20000000 when boot up.

2).Chainload by bootloader
If chainload by bootloader, the application must copy its vector table to SRAM at 0x20000000 so the bootloader's irq relay handler can find.

3).Boot directly.
As for boot directly, the vector table must put at address 0x0, no need to copy the vector table to SRAM since there is no bootloader (aka vector relay handler) to use.

Reply:
reply to 1) bootloader can have the relay function implemented in such a way that it only needs the pointer (call it <b>vr_pointer</b>) to current vector table. So bootloader will have one HW supported vector table at adress 0x00, and the 2nd one somewhere else (for instance just after the HW one). Initiali the pointer will point at the 2nd vector table.

reply to 2)
Application should have its vector table where it use to be placet normally. So it will be possibile to just update <b>vr_pointer</b> for relocate the vector. It is possible that this will be done by a) bootloader -just before it calls reset vector of application. b) in application reset vector.

What I described will decrease consumption of RAM for VR - usually M0 targets have not much of it.

Reply:
No,,, I don't want to implemented like that. Its, ugly I think ? I really don't like that way, use one flag to identify current irq is for bootloader or for application. Too bad.. too bad..

Only 192 byte, not too much. One single ISR will need at lease 600+ bytes for its stack.

nrf51822 QFAA have 16KB SRAM, the vector table in only occupied 1.1 percent of total SRAM.

Reply:
Let me guess, you may say: If you don't change it like what I said, your PR will never be merged into master branch. Hahaha

I will try it, I will @you when it's done. @nvlsianpu

Reply:
@miyatsu :

> If you don't change it like what I said, your PR will never be merged into master branch.

Let me chime in, as a mere developer, just you guys. I don't think that @nvlsianpu could say that, at least he doesn't have merge rights, just the same as you or me don't have. He can only either approve your PR, not approve, or -1. Generally, -1'ed PR wouldn't be merged by the maintainers with write access. You definitely could try to instead get approvals from other folks and then escalate merging even despite somebody's -1. But much more productive way is to listen to suggestions of other developers, and try to understand their point of view. Over time, as you get more experience with Zephyr, you may see that the suggestions were indeed useful/right.

Definitely please don't feel that you're "cornered" into making some change. The idea is to cooperate to write the *better* code than a lone developer could do themselves.

P.S. You're doing great with this PR and review, thanks for your contribution and patience, much appreciated!

Reply:
Oh, I'm just kidding. Sorry for that...

Reply:
@miyatsu - how cooperation looks like was described very good by @pfalcon.
You decide as contributor what you will develop, but indeed you need to get approval of your changes - not only from me. I was getting annoyed by heavy and long review many times. Functionality you try to add is extremely important to us (Nordic semiconductor I mean) and any cortex-M0 vendor - that why I em so demanding. (So Many times we were fighting for a few bytes of RAM more for the application)
BTW - I meant nota a switch, but the pointer.
Flexible vector relocation like this provided by CMSI void NVIC_SetVector(IRQn_Type IRQn, uint32_t vector) is another feature than expected here.